src/libraries/Multicurve.sol

```diff
@@ -0,0 +1,297 @@
+
+/// @notice Thrown when a curve has zero positions
+error ZeroPosition();
+
+/// @notice Thrown when a curve has zero share to be sold
+error ZeroShare();
+
+/// @notice Thrown when total shares are invalid (greater than WAD)
+error InvalidTotalShares();
+
+/**
+ * @dev Representation of a curve shape
+ * @param tickLower Lower tick of the curve
+ * @param tickUpper Upper tick of the curve
+ * @param numPositions Number of positions to create for this curve
+ * @param share Amount of shares to be sold on this curve (in WAD)
+ */
+struct Curve {
+    int24 tickLower;
+    int24 tickUpper;
+    uint16 numPositions;
+    uint256 shares;
+}
+
+/**
+ * @dev Adjusts and validates curves with an offset, and returns them along with the overall boundaries
+ * @param curves Array of curves to adjust and validate
+ * @param offset Offset to apply expressed in tick (`0` if no offset needed)
+ * @param tickSpacing Current tick spacing of the pool
+ * @param isToken0 True if the asset we're selling is token0, false otherwise
+ * @return adjustedCurves Array of adjusted and validated curves
+ * @return lowerTickBoundary Overall lower tick boundary across all curves
+ * @return upperTickBoundary Overall upper tick boundary across all curves
+ */
+function adjustCurves(
+    Curve[] memory curves,
+    int24 offset,
+    int24 tickSpacing,
+    bool isToken0
+) pure returns (Curve[] memory adjustedCurves, int24 lowerTickBoundary, int24 upperTickBoundary) {
+    uint256 length = curves.length;
+    adjustedCurves = new Curve[](length);
+
+    uint256 totalShares;
+
+    lowerTickBoundary = TickMath.MAX_TICK;
+    upperTickBoundary = TickMath.MIN_TICK;
+
+    for (uint256 i; i != length; ++i) {
+        Curve memory adjustedCurve = curves[i];
+
+        require(adjustedCurve.numPositions > 0, ZeroPosition());
+        require(adjustedCurve.shares > 0, ZeroShare());
+
+        if (offset != 0) {
+            adjustedCurve.tickLower += offset;
+            adjustedCurve.tickUpper += offset;
+        }
+
+        isTickAligned(adjustedCurve.tickLower, tickSpacing);
+        isTickAligned(adjustedCurve.tickUpper, tickSpacing);
+
+        // Flip the ticks if the asset is token1
+        if (!isToken0) {
+            adjustedCurve.tickLower = -adjustedCurve.tickUpper;
+            adjustedCurve.tickUpper = -adjustedCurve.tickLower;
+        }
+
+        isRangeOrdered(adjustedCurve.tickLower, adjustedCurve.tickUpper);
+
+        // Calculate the boundaries
+        if (lowerTickBoundary > adjustedCurve.tickLower) lowerTickBoundary = adjustedCurve.tickLower;
+        if (upperTickBoundary < adjustedCurve.tickUpper) upperTickBoundary = adjustedCurve.tickUpper;
+
+        // Accumulate the shares
+        totalShares += adjustedCurves[i].shares;
+
+        adjustedCurves[i] = adjustedCurve;
+    }
+
+    require(totalShares <= WAD, InvalidTotalShares());
+    // TODO: Might be an unnecessary check
+    isRangeOrdered(lowerTickBoundary, upperTickBoundary);
```

Comment:
Can confirm that this check is unnecessary since we validate that `tickLower[i] <= tickUpper[i]` so we can be certain that `min(tickLower) <= max(tickUpper)`